### ISSUE TITLE
Calculate distances in bulk and using numpy.

### ISSUE THREAD

**Issue by pmlandwehr** [geopy/geopy#189]
I was recently using Geopy to calculate distances of several hundred thousand lat longs from a single point. It was agonizingly slow - at least in the tens of minutes, if not hour+. Using an open haversine function in which I swapped out the default python math library for numpy and patched to allow for array processing, I got a speed up to ~ 1 minute. This would be a killer patch to get in.

**Comment by pmlandwehr** [comment]
To elaborate on this: it's more efficient to use `math` than `numpy` for single point operations. It's more efficient to use `numpy` than `math` for bulk operations:

```
In[1]: import numpy as np
In[2]: import math
In[3]: import numpy.random as npr
In[4]: degs  = [x+npr.random() for x in npr.randint(358, size=10000)]
In[5]: %timeit [np.radians(deg) for deg in degs]
Out[5]: 100 loops, best of 3: 6.43 ms per loop
In[6]: %timeit [math.radians(deg) for deg in degs]
Out[6]: 1000 loops, best of 3: 983 µs per loop
In[7]: %timeit np.radians(degs)
Out[7]: 1000 loops, best of 3: 551 µs per loop
```

**Comment by jqnatividad** [comment]
+1

**Comment by KostyaEsmukov** [comment]
I think that #144 should be implemented before we can start tackling with numpy, as it would make `vincenty` deprecated. But speeding up the new (geodesic) distance computation method might be better done by using `pyproj` instead of `numpy`. Discussion is here: https://github.com/geopy/geopy/pull/144#issuecomment-374047376

**Comment by cffk** [comment]
@pmlandwehr I know this goes beyond the question you asked; but you
might find it useful...

If the goal is to compute the closest distance of a point P to a
collection C of a `N` points, then you have to do `N` distance
calculations.  However, if you want to compute the closest point in C to
a bunch of `M` points, that you can much better than computing `MN`
distances by putting C into a VP tree.  The cost to construct the VP
tree is `N log N`; subsequently each closest distance calculation takes
`log N`.  So the total cost is `(N + M) log N`.

Lastly, I am interested at looking at implementing the geographiclib
distance function in numpy.  I did a similar "vectorized" implementation
for MATLAB.  Holding me up is the fact that I have no experience with
numpy!

**Comment by KostyaEsmukov** [comment]
> Holding me up is the fact that I have no experience with numpy!

Oh, I'm sure you'll figure this out easily. Let me know if I can help somehow.

**Comment by cffk** [comment]
Indeed...  The methodology needed is similar to that used for MATLAB.

**Comment by wboler05** [comment]
Hi, do we have an update on this?

**Comment by cffk** [comment]
I've made no progress on this, I'm afraid -- too many other irons in the fire.  It might help to know in more-or-less concrete terms what you application is.  How many points do have?  Do you need all possible O(N^2) distance calculations?  Or are you only interest in, for example, the closest neighbor to each point?

**Comment by drpepper19** [comment]
yes something similar to how sklearn calculates various distance measures between arrays. https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise_distances.html Or how scipy does it https://docs.scipy.org/doc/scipy/reference/spatial.distance.html

**Comment by cffk** [comment]
By the way, [pyproj](https://pyproj4.github.io/pyproj/stable) gives python access to distance calculations via the compiled C implementation.  This will be faster that the native python implementation.
